### ISSUE TITLE
Module defaults for k8s group using community.kubernetes collection

### ISSUE THREAD

**Issue by erikgb** [ansible-collections/kubernetes.core#126]
##### SUMMARY
I am trying to migrate our Ansible stuff to use FQN on module names (including migrating to `community.kubernetes` collection). Since we are using quite a few of the k8s modules in our playbooks, we have set module defaults for the k8s group, as explained [here](https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html#module-defaults-groups). I have made a few attempts to continue using module group defaults, but no luck so far.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
Seems to be a general issue with most modules in the `community.kubernetes` collection.

##### ANSIBLE VERSION

```paste below
ansible@abda10c4cd81:~$ ansible --version
ansible [core 2.11.1]
  config file = None
  configured module search path = ['/home/ansible/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.9/site-packages/ansible
  ansible collection location = /home/ansible/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.9.5 (default, May 12 2021, 15:26:36) [GCC 8.3.0]
  jinja version = 3.0.1
  libyaml = True
ansible@abda10c4cd81:~$
```

##### CONFIGURATION

```paste below
ansible@abda10c4cd81:~$ ansible-config dump --only-changed
ANSIBLE_FORCE_COLOR(env: ANSIBLE_FORCE_COLOR) = True
DEFAULT_LOAD_CALLBACK_PLUGINS(env: ANSIBLE_LOAD_CALLBACK_PLUGINS) = True
DEFAULT_STDOUT_CALLBACK(env: ANSIBLE_STDOUT_CALLBACK) = yaml
HOST_KEY_CHECKING(env: ANSIBLE_HOST_KEY_CHECKING) = False
INTERPRETER_PYTHON(env: ANSIBLE_PYTHON_INTERPRETER) = auto
ansible@abda10c4cd81:~$
```

##### OS / ENVIRONMENT

Container image based on official Python 3 image, running CI/CD with GitLab CI/CD.
````
ansible@abda10c4cd81:~$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
ansible@abda10c4cd81:~$
````

##### STEPS TO REPRODUCE

Working (current) configuration without FQN module name:
```yaml
- name: K8s Applications Provisioning
  hosts: k8s
  gather_facts: false
  module_defaults:
    group/k8s:
      api_key: "{{ kubernetes_api_key }}"
      host: "{{ kubernetes_host }}"
      namespace: "{{ kubernetes_namespace }}"
  tasks:
    - name: Get a list of all service accounts
      k8s_info:
        kind: ServiceAccount
```

Non-working configuration using FQN module name (community.kubernetes):
```yaml
- name: K8s Applications Provisioning
  hosts: k8s
  gather_facts: false
  module_defaults:
    group/k8s:
      api_key: "{{ kubernetes_api_key }}"
      host: "{{ kubernetes_host }}"
      namespace: "{{ kubernetes_namespace }}"
  tasks:
    - name: Get a list of all service accounts
      community.kubernetes.k8s_info:
        kind: ServiceAccount
```

I have tried various variants of specifying the group, but no luck so far (same result). E.g.

````yaml
  module_defaults:
    group/community.kubernetes:
````

##### EXPECTED RESULTS

````
TASK [Get a list of all service accounts] **************************************
ok: [REDACTED]
````

##### ACTUAL RESULTS

```paste below
TASK [Get a list of all service accounts] **************************************
fatal: [REDACTED]: FAILED! => changed=false 
  msg: Failed to load kubeconfig due to Invalid kube-config file. No configuration found.
```

**Comment by abikouo** [comment]
@erikgb thanks for taking the time to report this issue.
What you are trying to do is pending implementation of https://github.com/ansible/ansible/pull/74039, you can get full details here 
https://gist.github.com/s-hertel/725ecc719b5301e76c571aca58d39bd3.
For the time being you can update your playbook adding ``collections`` keyword like this:

```
- name: K8s Applications Provisioning
  hosts: k8s
  gather_facts: false
  
  collections:
    - community.kubernetes

  module_defaults:
    group/k8s:
      api_key: "{{ kubernetes_api_key }}"
      host: "{{ kubernetes_host }}"
      namespace: "{{ kubernetes_namespace }}"
  tasks:
    - name: Get a list of all service accounts
      k8s_info:
        kind: ServiceAccount
```

**Comment by erikgb** [comment]
@abikouo The blocking upstream PR seems merged now. Any chance to take a look at this now?

**Comment by gravesm** [comment]
@erikgb yes, we are working on it.

**Comment by gravesm** [comment]
Closing as this is now possible in Ansible 2.12.

**Comment by erikgb** [comment]
@gravesm I tested this now with Ansible 5 (ansible-core 2.12), and I get the following error:
````log
ERROR! Error loading module_defaults: could not resolve the module_defaults group ansible.builtin.kubernetes.core
````

I have tried the following syntax:
````yaml
module_defaults:
  group/kubernetes.core:
    (....)
````

Reading [the docs](https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html#module-defaults-groups), it seems like something has to be added to `meta/runtime.yml` in this collection to be able to use group defaults. WDYT? Reopen issue?

**Comment by gravesm** [comment]
Hi @erikgb, you have to specify the group name in your module defaults. There are currently two group names defined for this collection--`k8s` and `helm`. For example:

```yaml
module_defaults:
  group/kubernetes.core.k8s:
    namespace: foo
tasks:
  - kubernetes.core.k8s:
      definition:
        kind: ConfigMap
        apiVersion: v1
        metadata:
          name: myconfig
  - kubernetes.core.k8s_info:
      kind: ConfigMap
      name: myconfig
```

**Comment by erikgb** [comment]
@gravesm Thanks a lot for the explanation. I will give it another try. Thanks!

**Comment by erikgb** [comment]
Strangely, I now get this "funny" error message, but that might be related to something else:
````log
ERROR! [DEPRECATED]: kubernetes.core.k8s_facts has been removed. Use kubernetes.core.k8s_info instead. This feature was removed from kubernetes.core in version 2.0.0. Please update your playbooks.
````

I am not using the `k8s_facts` - as far as I know (and I have checked all dependencies), but I will check this in a simpler play and eventually file another issue. Any clues/hints?

**Comment by gravesm** [comment]
Ah, sorry. This bug has been fixed on main but not backported to the 2.2 branch. I'll try to get a new 2.2 release out next week with this fix.

**Comment by erikgb** [comment]
> Ah, sorry. This bug has been fixed on main but not backported to the 2.2 branch. I'll try to get a new 2.2 release out next week with this fix.

Ok, thanks! 🙏
